File: packages/zynth-crypto/ios/ZynthCryptoModule.swift
import CommonCrypto
import CryptoKit
import Foundation
import Security
import ZynthKit

@objc(ZynthCryptoModule)
final class ZynthCryptoModule: NSObject, ZynthModule, ZynthSyncModule {
  let name: String = "ZynthCrypto"

  var exportedMethods: [String] {
    [
      "getRandomBase64",
      "randomUUID",
      "digestBase64",
      "hmacBase64",
      "hkdfDeriveBitsBase64",
      "pbkdf2DeriveBitsBase64",
      "aesGcmEncryptBase64",
      "aesGcmDecryptBase64",
    ]
  }

  var protectedMethods: [String] {
    exportedMethods
  }

  func call(method: String, args: ZynthArgs) throws -> Any? {
    return try handle(method: method, args: args)
  }

  func callSync(method: String, args: ZynthArgs) throws -> Any? {
    return try handle(method: method, args: args)
  }

  private func handle(method: String, args: ZynthArgs) throws -> Any? {
    switch method {
    case "getRandomBase64":
      let size = Int(try args.number("size"))
      return ["result": try randomBytesBase64(size: size)]
    case "randomUUID":
      return ["result": UUID().uuidString.lowercased()]
    case "digestBase64":
      let algorithm = try args.string("algorithm")
      let dataBase64 = try args.string("dataBase64")
      return ["result": try digestBase64(algorithm: algorithm, dataBase64: dataBase64)]
    case "hmacBase64":
      let algorithm = try args.string("algorithm")
      let keyBase64 = try args.string("keyBase64")
      let dataBase64 = try args.string("dataBase64")
      return ["result": try hmacBase64(algorithm: algorithm, keyBase64: keyBase64, dataBase64: dataBase64)]
    case "hkdfDeriveBitsBase64":
      let algorithm = try args.string("algorithm")
      let keyBase64 = try args.string("keyBase64")
      let saltBase64 = try args.string("saltBase64")
      let infoBase64 = try args.string("infoBase64")
      let lengthBytes = Int(try args.number("lengthBytes"))
      return [
        "result": try hkdfDeriveBitsBase64(
          algorithm: algorithm,
          keyBase64: keyBase64,
          saltBase64: saltBase64,
          infoBase64: infoBase64,
          lengthBytes: lengthBytes
        )
      ]
    case "pbkdf2DeriveBitsBase64":
      let algorithm = try args.string("algorithm")
      let keyBase64 = try args.string("keyBase64")
      let saltBase64 = try args.string("saltBase64")
      let iterations = Int(try args.number("iterations"))
      let lengthBytes = Int(try args.number("lengthBytes"))
      return [
        "result": try pbkdf2DeriveBitsBase64(
          algorithm: algorithm,
          keyBase64: keyBase64,
          saltBase64: saltBase64,
          iterations: iterations,
          lengthBytes: lengthBytes
        )
      ]
    case "aesGcmEncryptBase64":
      return ["result": try aesGcmBase64(args: args, encrypt: true)]
    case "aesGcmDecryptBase64":
      return ["result": try aesGcmBase64(args: args, encrypt: false)]
    default:
      throw ZynthModuleError.methodNotExported(module: name, method: method)
    }
  }

  private func randomBytesBase64(size: Int) throws -> String {
    guard size >= 0 else {
      throw moduleError(code: 1, message: "size must be >= 0")
    }
    guard size <= 65_536 else {
      throw moduleError(code: 2, message: "size exceeds 65536 bytes")
    }
    if size == 0 {
      return ""
    }

    var bytes = [UInt8](repeating: 0, count: size)
    let status = SecRandomCopyBytes(kSecRandomDefault, size, &bytes)
    guard status == errSecSuccess else {
      throw moduleError(code: Int(status), message: "secure random generation failed")
    }
    return Data(bytes).base64EncodedString()
  }

  private func digestBase64(algorithm: String, dataBase64: String) throws -> String {
    let data = try decodeBase64(dataBase64, code: 3, message: "Invalid base64 input")
    guard data.count <= 16 * 1024 * 1024 else {
      throw moduleError(code: 4, message: "digest input exceeds maximum size")
    }

    let normalized = normalizeDigestName(algorithm)
    let digest: Data
    switch normalized {
    case "SHA-1":
      digest = Data(Insecure.SHA1.hash(data: data))
    case "SHA-256":
      digest = Data(SHA256.hash(data: data))
    case "SHA-384":
      digest = Data(SHA384.hash(data: data))
    case "SHA-512":
      digest = Data(SHA512.hash(data: data))
    default:
      throw moduleError(code: 5, message: "Unsupported digest algorithm")
    }

    return digest.base64EncodedString()
  }

  private func hmacBase64(algorithm: String, keyBase64: String, dataBase64: String) throws -> String {
    let keyData = try decodeBase64(keyBase64, code: 10, message: "Invalid key base64")
    let data = try decodeBase64(dataBase64, code: 11, message: "Invalid data base64")
    let normalized = normalizeDigestName(algorithm)

    let key = SymmetricKey(data: keyData)
    let macData: Data

    switch normalized {
    case "SHA-1":
      macData = Data(HMAC<Insecure.SHA1>.authenticationCode(for: data, using: key))
    case "SHA-256":
      macData = Data(HMAC<SHA256>.authenticationCode(for: data, using: key))
    case "SHA-384":
      macData = Data(HMAC<SHA384>.authenticationCode(for: data, using: key))
    case "SHA-512":
      macData = Data(HMAC<SHA512>.authenticationCode(for: data, using: key))
    default:
      throw moduleError(code: 12, message: "Unsupported HMAC algorithm")
    }

    return macData.base64EncodedString()
  }

  private func hkdfDeriveBitsBase64(
    algorithm: String,
    keyBase64: String,
    saltBase64: String,
    infoBase64: String,
    lengthBytes: Int
  ) throws -> String {
    guard lengthBytes > 0 else {
      throw moduleError(code: 13, message: "lengthBytes must be > 0")
    }
    guard lengthBytes <= 8192 else {
      throw moduleError(code: 14, message: "lengthBytes exceeds 8192")
    }

    let keyData = try decodeBase64(keyBase64, code: 15, message: "Invalid key base64")
    let saltData = try decodeBase64(saltBase64, code: 16, message: "Invalid salt base64")
    let infoData = try decodeBase64(infoBase64, code: 17, message: "Invalid info base64")

    let normalized = normalizeDigestName(algorithm)
    let inputKey = SymmetricKey(data: keyData)

    let derived: SymmetricKey
    switch normalized {
    case "SHA-1":
      derived = HKDF<Insecure.SHA1>.deriveKey(
        inputKeyMaterial: inputKey,
        salt: saltData,
        info: infoData,
        outputByteCount: lengthBytes
      )
    case "SHA-256":
      derived = HKDF<SHA256>.deriveKey(
        inputKeyMaterial: inputKey,
        salt: saltData,
        info: infoData,
        outputByteCount: lengthBytes
      )
    case "SHA-384":
      derived = HKDF<SHA384>.deriveKey(
        inputKeyMaterial: inputKey,
        salt: saltData,
        info: infoData,
        outputByteCount: lengthBytes
      )
    case "SHA-512":
      derived = HKDF<SHA512>.deriveKey(
        inputKeyMaterial: inputKey,
        salt: saltData,
        info: infoData,
        outputByteCount: lengthBytes
      )
    default:
      throw moduleError(code: 18, message: "Unsupported HKDF algorithm")
    }

    let bytes = derived.withUnsafeBytes { Data($0) }
    return bytes.base64EncodedString()
  }

  private func pbkdf2DeriveBitsBase64(
    algorithm: String,
    keyBase64: String,
    saltBase64: String,
    iterations: Int,
    lengthBytes: Int
  ) throws -> String {
    guard iterations > 0 else {
      throw moduleError(code: 19, message: "iterations must be > 0")
    }
    guard lengthBytes > 0 else {
      throw moduleError(code: 20, message: "lengthBytes must be > 0")
    }
    guard lengthBytes <= 8192 else {
      throw moduleError(code: 21, message: "lengthBytes exceeds 8192")
    }

    let keyData = try decodeBase64(keyBase64, code: 22, message: "Invalid key base64")
    let saltData = try decodeBase64(saltBase64, code: 23, message: "Invalid salt base64")

    let normalized = normalizeDigestName(algorithm)
    let prf: CCPseudoRandomAlgorithm
    switch normalized {
    case "SHA-1":
      prf = CCPseudoRandomAlgorithm(kCCPRFHmacAlgSHA1)
    case "SHA-256":
      prf = CCPseudoRandomAlgorithm(kCCPRFHmacAlgSHA256)
    case "SHA-384":
      prf = CCPseudoRandomAlgorithm(kCCPRFHmacAlgSHA384)
    case "SHA-512":
      prf = CCPseudoRandomAlgorithm(kCCPRFHmacAlgSHA512)
    default:
      throw moduleError(code: 24, message: "Unsupported PBKDF2 algorithm")
    }

    var output = [UInt8](repeating: 0, count: lengthBytes)
    let status = keyData.withUnsafeBytes { keyPtr in
      saltData.withUnsafeBytes { saltPtr in
        CCKeyDerivationPBKDF(
          CCPBKDFAlgorithm(kCCPBKDF2),
          keyPtr.bindMemory(to: Int8.self).baseAddress,
          keyData.count,
          saltPtr.bindMemory(to: UInt8.self).baseAddress,
          saltData.count,
          prf,
          UInt32(iterations),
          &output,
          lengthBytes
        )
      }
    }

    guard status == kCCSuccess else {
      throw moduleError(code: 25, message: "PBKDF2 derivation failed")
    }

    return Data(output).base64EncodedString()
  }

  private func aesGcmBase64(args: ZynthArgs, encrypt: Bool) throws -> String {
    let keyData = try decodeBase64(args.string("keyBase64"), code: 26, message: "Invalid key base64")
    let ivData = try decodeBase64(args.string("ivBase64"), code: 27, message: "Invalid iv base64")
    let additionalData = try decodeBase64(
      args.string("additionalDataBase64", default: ""),
      code: 28,
      message: "Invalid additionalData base64"
    )
    let data = try decodeBase64(args.string("dataBase64"), code: 29, message: "Invalid data base64")
    let tagLengthBits = Int(try args.number("tagLengthBits"))

    guard tagLengthBits == 128 else {
      throw moduleError(code: 30, message: "AES-GCM supports tagLength 128 bits only")
    }
    guard ivData.count >= 12 && ivData.count <= 16 else {
      throw moduleError(code: 31, message: "AES-GCM iv length must be between 12 and 16 bytes")
    }

    let keyBitLength = keyData.count * 8
    guard keyBitLength == 128 || keyBitLength == 192 || keyBitLength == 256 else {
      throw moduleError(code: 32, message: "AES-GCM key length must be 128, 192, or 256 bits")
    }

    let key = SymmetricKey(data: keyData)
    let nonce = try AES.GCM.Nonce(data: ivData)

    if encrypt {
      let sealed = try AES.GCM.seal(data, using: key, nonce: nonce, authenticating: additionalData)
      guard let combined = sealed.combined else {
        throw moduleError(code: 33, message: "AES-GCM encryption failed")
      }
      return combined.base64EncodedString()
    }

    let box = try AES.GCM.SealedBox(combined: data)
    let decrypted = try AES.GCM.open(box, using: key, authenticating: additionalData)
    return decrypted.base64EncodedString()
  }

  private func decodeBase64(_ value: String, code: Int, message: String) throws -> Data {
    if value.isEmpty {
      return Data()
    }
    guard let decoded = Data(base64Encoded: value) else {
      throw moduleError(code: code, message: message)
    }
    return decoded
  }

  private func normalizeDigestName(_ algorithm: String) -> String {
    algorithm.trimmingCharacters(in: .whitespacesAndNewlines).uppercased()
  }

  private func moduleError(code: Int, message: String) -> NSError {
    NSError(domain: "ZynthCrypto", code: code, userInfo: [NSLocalizedDescriptionKey: message])
  }
}

File: packages/zynth-crypto/android/ZynthCrypto/src/main/java/dev/zynth/crypto/ZynthCryptoModule.kt
package dev.zynth.crypto

import android.util.Base64
import com.zynth.kit.runtime.ZynthArgs
import com.zynth.kit.runtime.ZynthModule
import com.zynth.kit.runtime.ZynthSyncModule
import org.json.JSONObject
import java.security.MessageDigest
import java.security.SecureRandom
import java.util.Locale
import java.util.UUID
import javax.crypto.Cipher
import javax.crypto.Mac
import javax.crypto.spec.GCMParameterSpec
import javax.crypto.spec.SecretKeySpec
import kotlin.math.ceil

class ZynthCryptoModule : ZynthModule, ZynthSyncModule {
  override val name: String = "ZynthCrypto"

  override val exportedMethods: List<String> = listOf(
    "getRandomBase64",
    "randomUUID",
    "digestBase64",
    "hmacBase64",
    "hkdfDeriveBitsBase64",
    "pbkdf2DeriveBitsBase64",
    "aesGcmEncryptBase64",
    "aesGcmDecryptBase64",
  )

  override val protectedMethods: List<String> = exportedMethods

  private val secureRandom = SecureRandom()

  override fun call(method: String, args: ZynthArgs): JSONObject {
    return resultResponse(callSync(method, args))
  }

  override fun callSync(method: String, args: ZynthArgs): Any? {
    return when (method) {
      "getRandomBase64" -> randomBytesBase64(args.getInt("size"))
      "randomUUID" -> UUID.randomUUID().toString()
      "digestBase64" -> digestBase64(args.getString("algorithm"), args.getString("dataBase64"))
      "hmacBase64" -> hmacBase64(
        args.getString("algorithm"),
        args.getString("keyBase64"),
        args.getString("dataBase64")
      )
      "hkdfDeriveBitsBase64" -> hkdfDeriveBitsBase64(
        args.getString("algorithm"),
        args.getString("keyBase64"),
        args.getString("saltBase64"),
        args.getString("infoBase64"),
        args.getInt("lengthBytes")
      )
      "pbkdf2DeriveBitsBase64" -> pbkdf2DeriveBitsBase64(
        args.getString("algorithm"),
        args.getString("keyBase64"),
        args.getString("saltBase64"),
        args.getInt("iterations"),
        args.getInt("lengthBytes")
      )
      "aesGcmEncryptBase64" -> aesGcmBase64(args, encrypt = true)
      "aesGcmDecryptBase64" -> aesGcmBase64(args, encrypt = false)
      else -> throw IllegalArgumentException("Unsupported method: $method")
    }
  }

  private fun randomBytesBase64(size: Int): String {
    require(size >= 0) { "size must be >= 0" }
    require(size <= 65_536) { "size exceeds 65536 bytes" }
    if (size == 0) {
      return ""
    }

    val bytes = ByteArray(size)
    secureRandom.nextBytes(bytes)
    return toBase64(bytes)
  }

  private fun digestBase64(algorithm: String, dataBase64: String): String {
    val input = fromBase64(dataBase64)
    require(input.size <= 16 * 1024 * 1024) { "digest input exceeds maximum size" }

    val digestAlgorithm = normalizeDigestAlgorithm(algorithm)
    val digest = MessageDigest.getInstance(digestAlgorithm).digest(input)
    return toBase64(digest)
  }

  private fun hmacBase64(algorithm: String, keyBase64: String, dataBase64: String): String {
    val key = fromBase64(keyBase64)
    val data = fromBase64(dataBase64)
    val mac = Mac.getInstance(toHmacAlgorithm(algorithm))
    mac.init(SecretKeySpec(key, "RAW"))
    return toBase64(mac.doFinal(data))
  }

  private fun hkdfDeriveBitsBase64(
    algorithm: String,
    keyBase64: String,
    saltBase64: String,
    infoBase64: String,
    lengthBytes: Int
  ): String {
    require(lengthBytes > 0) { "lengthBytes must be > 0" }
    require(lengthBytes <= 8192) { "lengthBytes exceeds 8192" }

    val ikm = fromBase64(keyBase64)
    val salt = fromBase64(saltBase64)
    val info = fromBase64(infoBase64)

    val hmacAlgo = toHmacAlgorithm(algorithm)
    val hashLength = digestLength(normalizeDigestAlgorithm(algorithm))
    val prk = hmac(hmacAlgo, if (salt.isNotEmpty()) salt else ByteArray(hashLength), ikm)

    val blocks = ceil(lengthBytes.toDouble() / hashLength.toDouble()).toInt()
    require(blocks <= 255) { "requested HKDF output too large" }

    var previous = ByteArray(0)
    val output = ByteArray(lengthBytes)
    var offset = 0

    for (index in 1..blocks) {
      val context = ByteArray(previous.size + info.size + 1)
      System.arraycopy(previous, 0, context, 0, previous.size)
      System.arraycopy(info, 0, context, previous.size, info.size)
      context[context.lastIndex] = index.toByte()

      previous = hmac(hmacAlgo, prk, context)
      val copyLength = minOf(previous.size, lengthBytes - offset)
      System.arraycopy(previous, 0, output, offset, copyLength)
      offset += copyLength
    }

    return toBase64(output)
  }

  private fun pbkdf2DeriveBitsBase64(
    algorithm: String,
    keyBase64: String,
    saltBase64: String,
    iterations: Int,
    lengthBytes: Int
  ): String {
    require(iterations > 0) { "iterations must be > 0" }
    require(lengthBytes > 0) { "lengthBytes must be > 0" }
    require(lengthBytes <= 8192) { "lengthBytes exceeds 8192" }

    val password = fromBase64(keyBase64)
    val salt = fromBase64(saltBase64)
    val hmacAlgo = toHmacAlgorithm(algorithm)
    val hashLen = digestLength(normalizeDigestAlgorithm(algorithm))

    val blocks = ceil(lengthBytes.toDouble() / hashLen.toDouble()).toInt()
    val output = ByteArray(lengthBytes)
    var outputOffset = 0

    for (block in 1..blocks) {
      val saltBlock = ByteArray(salt.size + 4)
      System.arraycopy(salt, 0, saltBlock, 0, salt.size)
      saltBlock[salt.size] = ((block ushr 24) and 0xff).toByte()
      saltBlock[salt.size + 1] = ((block ushr 16) and 0xff).toByte()
      saltBlock[salt.size + 2] = ((block ushr 8) and 0xff).toByte()
      saltBlock[salt.size + 3] = (block and 0xff).toByte()

      var u = hmac(hmacAlgo, password, saltBlock)
      val t = u.copyOf()

      for (round in 2..iterations) {
        u = hmac(hmacAlgo, password, u)
        for (index in t.indices) {
          t[index] = (t[index].toInt() xor u[index].toInt()).toByte()
        }
      }

      val copyLength = minOf(t.size, lengthBytes - outputOffset)
      System.arraycopy(t, 0, output, outputOffset, copyLength)
      outputOffset += copyLength
    }

    return toBase64(output)
  }

  private fun aesGcmBase64(args: ZynthArgs, encrypt: Boolean): String {
    val key = fromBase64(args.getString("keyBase64"))
    val iv = fromBase64(args.getString("ivBase64"))
    val additionalData = fromBase64(args.getString("additionalDataBase64", ""))
    val data = fromBase64(args.getString("dataBase64"))
    val tagLengthBits = args.getInt("tagLengthBits")

    require(tagLengthBits == 128) { "AES-GCM supports tagLength 128 bits only" }
    require(iv.size in 12..16) { "AES-GCM iv length must be between 12 and 16 bytes" }

    val keyBits = key.size * 8
    require(keyBits == 128 || keyBits == 192 || keyBits == 256) {
      "AES-GCM key length must be 128, 192, or 256 bits"
    }

    val cipher = Cipher.getInstance("AES/GCM/NoPadding")
    val keySpec = SecretKeySpec(key, "AES")
    val gcmSpec = GCMParameterSpec(tagLengthBits, iv)

    if (encrypt) {
      cipher.init(Cipher.ENCRYPT_MODE, keySpec, gcmSpec)
    } else {
      cipher.init(Cipher.DECRYPT_MODE, keySpec, gcmSpec)
    }

    if (additionalData.isNotEmpty()) {
      cipher.updateAAD(additionalData)
    }

    val output = cipher.doFinal(data)
    return toBase64(output)
  }

  private fun normalizeDigestAlgorithm(algorithm: String): String {
    return when (algorithm.trim().uppercase(Locale.ROOT)) {
      "SHA-1" -> "SHA-1"
      "SHA-256" -> "SHA-256"
      "SHA-384" -> "SHA-384"
      "SHA-512" -> "SHA-512"
      else -> throw IllegalArgumentException("Unsupported digest algorithm")
    }
  }

  private fun digestLength(algorithm: String): Int {
    return when (algorithm) {
      "SHA-1" -> 20
      "SHA-256" -> 32
      "SHA-384" -> 48
      "SHA-512" -> 64
      else -> throw IllegalArgumentException("Unsupported digest algorithm")
    }
  }

  private fun toHmacAlgorithm(algorithm: String): String {
    return when (normalizeDigestAlgorithm(algorithm)) {
      "SHA-1" -> "HmacSHA1"
      "SHA-256" -> "HmacSHA256"
      "SHA-384" -> "HmacSHA384"
      "SHA-512" -> "HmacSHA512"
      else -> throw IllegalArgumentException("Unsupported HMAC algorithm")
    }
  }

  private fun hmac(algorithm: String, key: ByteArray, data: ByteArray): ByteArray {
    val mac = Mac.getInstance(algorithm)
    mac.init(SecretKeySpec(key, "RAW"))
    return mac.doFinal(data)
  }

  private fun toBase64(data: ByteArray): String {
    return Base64.encodeToString(data, Base64.NO_WRAP)
  }

  private fun fromBase64(data: String): ByteArray {
    if (data.isEmpty()) {
      return ByteArray(0)
    }
    return try {
      Base64.decode(data, Base64.DEFAULT)
    } catch (_: IllegalArgumentException) {
      throw IllegalArgumentException("Invalid base64 input")
    }
  }

  private fun resultResponse(result: Any?): JSONObject {
    return JSONObject().apply {
      put("result", result ?: JSONObject.NULL)
    }
  }
}
